rfc-108-including-gem-component-assets.md

```diff
@@ -0,0 +1,111 @@
+
+In the first instance we could search through the code base and write a test for that application to ensure the right assets are being included. This test would need to be kept up to date as components are added and removed.
+
+We could also look at altering the component guide to either mark components in use or put them in a separate list. That way we could look at the component guide for an application and immediately know which components are in use and which are not.
+
+## Alternatives
+
+### Caching on GOV.UK
+
+The way that the Sass is currently included in each application makes caching ineffective. When a user lands on a page on GOV.UK, their browser downloads the CSS for that application (say `government-frontend`). This is a single file that includes custom styles from that application plus the styles from the gem. If they then navigate to a page served by another application (say `collections`) they download a new CSS file that contains the exact same styles from the gem, but because it is mixed in with the custom styles for that application, they are different files and cannot be cached.
+
+The solution above does not solve this problem, but it would potentially reduce the size of each downloaded CSS file. An optimal solution would be to separate the current `@import 'govuk_publishing_components/all_components';` statement into a separate stylesheet for each application. This would add another stylesheet, but make it cachable. The structure of each application's stylesheets would look something like this:
+
+```
+// _application.scss
+.styles_for_this_application {
+  ...
+}
+```
+
+```
+// _components.scss
+@import 'govuk_publishing_components/all_components';
+```
+
+The source of a page on GOV.UK would look something like this:
+
+```
+<link rel="stylesheet" media="screen" href="https://assets.publishing.service.gov.uk/frontend/application-a6d734239cc6bfc2534ca11a838575ef7b.css" crossorigin="anonymous" integrity="sha256-73t28PTeOrcJZQJGEPkS9ww=">
+
+<link rel="stylesheet" media="screen" href="https://assets.publishing.service.gov.uk/frontend/components-76f0f4de3ab70965024610f912f70c.css" crossorigin="anonymous" integrity="sha256-ptc0I5zGv8JTTKEag4V1=">
+```
+
+The [rails fingerprinting](https://guides.rubyonrails.org/asset_pipeline.html#what-is-fingerprinting-and-why-should-i-care-questionmark) would ensure that the file is cached between applications, based on its content and size (not date).
+
+This would only work if the components stylesheet in each application was the same, so the solution described above of only including the components needed would not work. Instead, we would need to separate out the public facing component styles from the private application styles. For example (in the gem):
```

Comment:
Not sure if I'm reading this right - would this mean we wouldn't be choosing which components to include in each app, but just choosing whether to include public or private components?

Reply:
Yes.